### DISCUSSION TITLE
Table: Document possibility to use function as dataKey

### DISCUSSION THREAD

**Discussion by perceptron8** [primefaces/community#3248] [PrimeNG]
`Table`'s `dataKey` is currently (v19.0.0) declared as:

```ts
@Input() dataKey: string | undefined;
```

As under the hood `ObjectUtils.resolveFieldData` is being used, it is possible to pass a function instead of string. It is, however, not documented and not guaranteed to work.

It would be great if `dataKey` was declared as:

```ts
@Input() dataKey: string | Function | undefined;
```

In our use case, we are iterating over `SomeEnum` constants and then do row expansion using `Map<SomeEnum, BoringDetails[]>`. Sadly, it's impossible to use `rowexpansion` without a `dataKey` (expansion leads to "Error: dataKey or groupRowsBy must be defined to use row expansion"), yet... enum doesn't have any `dataKey` that could be used (however it is _itself_ perfectly valid `dataKey`!).

We would happily use `dataKey="identity"` if that was possible (assuming `identity` being equal `v => v`) .
Currently we are forced to use `dataKey="$any(identity)"` and pray that the implementation doesn't change. 

See [#13196](https://github.com/primefaces/primeng/issues/13196).

**Comment by stijnherreman** [comment]
Related: https://github.com/primefaces/primeng/issues/5383
A comment from 2020 demonstrates how to use a function as `dataKey`, but I'm running into the same typing issue. `$any()` doesn't work, although that's probably because we have a custom component wrapping the Table component.
Either way, taking a dependency on (what seems to be) an implementation detail does not feel right.
